File: tests/test_profile_lever.py
"""The strategy profile's runtime lever, Python half (2026-07-23).

The SHIPPED config carries the SHIPPED profile (swing). The runtime choice
lives in controls.json ("strategy_profile"), resolved by
market_data.universe.resolved_profile the same way core/profile_controls.hpp
resolves it for the engine, so both halves read the same value. An invalid or
unreadable control file means NO override, config decides.

The first test is the guard against the old mechanism returning: a session
that reintroduces a profile edit to config/default_config.yaml fails it,
committed or not, because it reads the working-tree file.
"""
from __future__ import annotations

import json
import os

SHIPPED = "config/default_config.yaml"


def _write_controls(tmp_path, payload: dict) -> None:
    with open(os.path.join(str(tmp_path), "controls.json"), "w") as fh:
        json.dump(payload, fh)


def test_shipped_config_carries_the_shipped_profile():
    from llm_consensus.config_access import config_block
    profile = str(config_block("strategy", SHIPPED).get("profile", "swing"))
    assert profile == "swing", (
        f"config/default_config.yaml ships profile '{profile}'. The shipped "
        "default is swing; select active_quant through the controls.json "
        "strategy_profile lever, never by editing the shipped file (that "
        "edit got swept into commit 440fda8)")


def test_control_file_overrides_the_profile(tmp_path, monkeypatch):
    monkeypatch.setenv("MAL_CONTROL_DIR", str(tmp_path))
    _write_controls(tmp_path, {"strategy_profile": "active_quant"})
    from market_data import universe
    assert universe.resolved_profile(SHIPPED) == "active_quant"
    core = universe.declared_core(SHIPPED)
    assert "SOL/USD" in core and len(core) == 8, (
        "the active_quant override must resolve the eight-name core")


def test_invalid_or_absent_override_means_config_decides(tmp_path, monkeypatch):
    from market_data import universe
    monkeypatch.setenv("MAL_CONTROL_DIR", str(tmp_path))
    # Absent file: config decides.
    assert universe.resolved_profile(SHIPPED) == "swing"
    # Invalid value: refused, never guessed.
    _write_controls(tmp_path, {"strategy_profile": "yolo_mode"})
    assert universe.resolved_profile(SHIPPED) == "swing"
    core = universe.declared_core(SHIPPED)
    assert len(core) == 4, "an invalid override must leave the swing core"


def test_controls_writer_round_trips_the_profile(tmp_path, monkeypatch):
    monkeypatch.setenv("MAL_CONTROL_DIR", str(tmp_path))
    from api_server import controls
    state = controls.read_controls()
    assert state.get("strategy_profile") in controls.STRATEGY_PROFILES
    state["strategy_profile"] = "active_quant"
    controls._write_controls(state)
    again = controls.read_controls()
    assert again["strategy_profile"] == "active_quant"
    # Flat and unique in the written file: the engine's reader is a flat
    # search, so a duplicate key would be read order-dependently.
    text = open(os.path.join(str(tmp_path), "controls.json")).read()
    assert text.count('"strategy_profile"') == 1

File: tests/test_profile_lever.cpp
// The strategy profile's runtime lever, C++ half (2026-07-23).
//
// The SHIPPED config carries the SHIPPED profile (swing): the runtime choice
// lives in controls.json ("strategy_profile"), never as an edit to the
// shipped default (that edit got swept into commit 440fda8, which is the
// failure mode this lever removes). The override is applied at load, so the
// active_quant overlay keys off the RESOLVED profile; an invalid or missing
// value means NO override, config decides.
#include <cstdio>
#include <fstream>
#include <string>

#include "config/config.hpp"
#include "core/profile_controls.hpp"
#include "test_util.hpp"

using namespace mal;

namespace {
void write_file(const std::string& path, const std::string& body) {
    std::ofstream out(path);
    out << body;
}
}  // namespace

int main() {
    // The shipped file ships the shipped profile. This is the guard that
    // fails if a future session reintroduces a profile edit to the yaml as
    // the mechanism instead of the control-file lever.
    config::Config shipped = config::load_config("config/default_config.yaml");
    maltest::check(shipped.strategy.profile == "swing",
                   "the SHIPPED config carries the SHIPPED profile (swing); "
                   "select active_quant through controls.json "
                   "strategy_profile, never by editing the shipped default");
    maltest::check(shipped.strategy.whitelist.size() == 4,
                   "shipped (swing) whitelist is the four-name core");

    // The override applies the overlay at load.
    config::Config aq =
        config::load_config("config/default_config.yaml", "active_quant");
    maltest::check(aq.strategy.profile == "active_quant",
                   "the override selects active_quant");
    maltest::check(aq.strategy.reversion_style == "rsi2",
                   "the active_quant overlay applied (rsi2 reversion)");
    maltest::check(aq.strategy.whitelist.size() == 8,
                   "the active_quant overlay applied (eight-name core)");

    // The resolver: valid value wins, invalid and missing mean no override.
    const std::string dir = "/tmp";
    const std::string ok = dir + "/mal_test_profile_ok.json";
    const std::string bad = dir + "/mal_test_profile_bad.json";
    const std::string missing = dir + "/mal_test_profile_missing.json";
    write_file(ok, "{\"strategy_profile\": \"active_quant\", \"x\": 1}");
    write_file(bad, "{\"strategy_profile\": \"yolo_mode\"}");
    std::remove(missing.c_str());
    maltest::check(core::resolve_profile_override(ok) == "active_quant",
                   "a valid control-file profile overrides");
    maltest::check(core::resolve_profile_override(bad).empty(),
                   "an invalid profile value is refused, never guessed");
    maltest::check(core::resolve_profile_override(missing).empty(),
                   "a missing control file means no override, config decides");
    std::remove(ok.c_str());
    std::remove(bad.c_str());

    return maltest::report("profile_lever");
}
